myapp/views.py

```diff
@@ -59,85 +60,85 @@ def single_stock_historic(request, symbol):
-                    t.save()
-                    user_profile.balance -= price
-                    if stock_symbol in user_profile.stocks:
-                        user_profile.stocks[stock_symbol] += number_of_stocks
-                    else:
-                        user_profile.stocks[stock_symbol] = number_of_stocks
-                    user_profile.save()
-                    content.update({'success': f'You have bought {number_of_stocks} {stock_symbol} for a price {price}'
-                                               f'\n your current balance is: {user_profile.balance}'})
-
-                else:
-                    content.update({"error": "not enough balance"})
-                    return render(request, 'trade.html', content)
-
-            elif 'sell' in params:
-                available_stocks = models.UserProfile.objects.get(user=request.user).stocks[stock_symbol]
-                if available_stocks >= number_of_stocks:
-                    t = models.Transaction.objects.create(user_id=models.UserProfile.objects.get(user=request.user),
-                                                          stock_symbol=stock_symbol,
-                                                          trans_date=datetime.now(),
-                                                          buy_or_sell=1,
-                                                          quantity=number_of_stocks,
-                                                          price=price)
-                    t.save()
-
-                    user_profile.stocks[stock_symbol] -= number_of_stocks
-
-                    user_profile.save()
-
-                    content.update({'success': f'You have sold {number_of_stocks} for a price {price}'
-                                               f'\n your current balance is: {user_profile.balance}'})
-                else:
-                    content.update({"error": "not enough stocks"})
-        else:
-            return redirect('login')
+    content = {'stock_list': stock_list}
+    rendered_page = render(request, 'trade.html', content)
+    return rendered_page
 
 
+@login_required
+@csrf_exempt
+def buy(request):
+    stock_list = stock_api.get_all_stocks()
+    context = {'stock_list': stock_list}
+    if request.method == 'POST':
+        params = request.POST
 
+        stock_info = trade_logic.extract_info(params)
 
-    rendered_page = render(request, 'trade.html', content)
-    return rendered_page
+        if stock_info is None:
+            context.update({"error": "Invalid input"})
+            return render(request, 'trade.html', context)
 
+        user_profile = models.UserProfile.objects.get(user=request.user)
+
+        if stock_info['total_price'] > user_profile.balance:
+            context.update({"error": "not enough balance"})
+            return render(request, 'trade.html', context)
+
+        user_profile.balance -= stock_info['total_price']
+        user_profile.save()

```

Comment:
you are updating the balance, before adding the transaaction
if the transactoin fails, you will have a very hard to debug problem.